apps/serdes/Serializer.cpp

```diff
@@ -0,0 +1,755 @@
+    // bits
+    int bits = type.bits();
+
+    // lanes
+    int lanes = type.lanes();
+
+    // code
+    halide_type_code_t code = type.code();
+
+    auto code_serialized = Halide::Serdes::TypeCode(code);
+
+    auto type_obj = Halide::Serdes::CreateType(builder, code_serialized, bits, lanes);
+    return type_obj;
+}
+
+std::pair<Halide::Serdes::Stmt, flatbuffers::Offset<void>> Serializer::serialize_stmt(flatbuffers::FlatBufferBuilder &builder, const Halide::Internal::Stmt &stmt) {
+    switch (stmt->node_type) {
+    case Halide::Internal::IRNodeType::LetStmt: {
+        auto let_stmt = stmt.as<Halide::Internal::LetStmt>();
+        std::string name = let_stmt->name;
+        auto name_serialized = serialize_string(builder, name);
+        auto value = let_stmt->value;
+        auto value_serialized = serialize_expr(builder, value);
+        auto body = let_stmt->body;
+        auto body_serialized = serialize_stmt(builder, body);
+        return std::make_pair(Halide::Serdes::Stmt::Stmt_LetStmt, Halide::Serdes::CreateLetStmt(builder, name_serialized, value_serialized.first, value_serialized.second, body_serialized.first, body_serialized.second).Union());
+    }
+    case Halide::Internal::IRNodeType::AssertStmt: {
+        auto assert_stmt = stmt.as<Halide::Internal::AssertStmt>();
+        auto condition = assert_stmt->condition;
+        auto condition_serialized = serialize_expr(builder, condition);
+        auto message = assert_stmt->message;
+        auto message_serialized = serialize_expr(builder, message);
+        return std::make_pair(Halide::Serdes::Stmt::Stmt_AssertStmt, Halide::Serdes::CreateAssertStmt(builder, condition_serialized.first, condition_serialized.second, message_serialized.first, message_serialized.second).Union());
+    }
+    case Halide::Internal::IRNodeType::ProducerConsumer: {
+        auto producer_consumer = stmt.as<Halide::Internal::ProducerConsumer>();
+        std::string name = producer_consumer->name;
+        auto name_serialized = serialize_string(builder, name);
+        bool is_producer = producer_consumer->is_producer;
+        auto body = producer_consumer->body;
+        auto body_serialized = serialize_stmt(builder, body);
+        return std::make_pair(Halide::Serdes::Stmt::Stmt_ProducerConsumer, Halide::Serdes::CreateProducerConsumer(builder, name_serialized, is_producer, body_serialized.first, body_serialized.second).Union());
+    }
+    case Halide::Internal::IRNodeType::For: {
+        auto for_stmt = stmt.as<Halide::Internal::For>();
+        std::string name = for_stmt->name;
+        auto name_serialized = serialize_string(builder, name);
+        auto min = for_stmt->min;
+        auto min_serialized = serialize_expr(builder, min);
+        auto extent = for_stmt->extent;
+        auto extent_serialized = serialize_expr(builder, extent);
+        Halide::Serdes::ForType for_type = Halide::Serdes::ForType::ForType_Serial;
+        switch (for_stmt->for_type) {
```

Comment:
(Here and elsewhere:)

Using a switch statement like this to ensure that the serialized values match the compiled-into-code values is absolutely the safest thing to do to ensure we don't end up with embarrassing breakages if these get out of sync.

That said, as a bit of premature optimization... what if we did something like

```
static_assert((int)Halide::Internal::ForType::Serial == (int)Halide::Serdes::ForType::ForType_Serial);
static_assert((int)Halide::Internal::ForType::Parallel == (int)Halide::Serdes::ForType::ForType_Parallel);
... etc for the rest
for_type = (Halide::Serdes::ForType)(int)for_stmt->for_type;
```

...ewww. After having typed that out, it's definitely ugly and I'd avoid it unless I had profiling data that showed it was a meaningful performance improvement.